### ISSUE TITLE
Tests stops at "mocha:test"

### ISSUE THREAD

**Issue by niksy** [Modernizr/Modernizr#2355]
When running on Node 8, tests stop at `mocha:test` task.

```sh
> grunt test --stack

(node:4106) ExperimentalWarning: The http2 module is an experimental API.
Running "clean:dist" (clean) task
>> 5 paths cleaned.

Running "eslint:target" (eslint) task

Running "pug:compile" (pug) task
>> 3 files created.

Running "instrument:files" (instrument) task
Instrumented 53 files

Running "env:coverage" (env) task

Running "mochaTest:test" (mochaTest) task


  ․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․
  ․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․
  ․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․
  ․․․․․․․․․․․․․․․․․․․․

  323 passing (2s)


Running "generate:dest" (generate) task

Running "connect:server" (connect) task
Started connect web server on http://localhost:9090

Running "mocha:test" (mocha) task
Testing: http://localhost:9090/test/unit.html
```

**Comment by niksy** [comment]
@rejas When I update grunt-mocha to latest version, I get proper output but a lot of errors. I suppose setting larger version of grunt-mocha isn’t desirable right now?

What about using different runner for browser tests? For example, Karma is standard for these things so maybe we could use it? Also, PhantomJS is deprectaed and I didn’t see any support for using another modern headless solution like Puppeteer with Gunt, but karma-chrome-launcher has support for that.

**Comment by rejas** [comment]
@niksy yeah, the test suite is not really upgrade friendly (tried to update [mocha](https://github.com/Modernizr/Modernizr/tree/update_grunt-mocha) like you did with similar results). grunt-mocha isnt also very responsive to PRs. 
So I'd also like to see a refresh in that area but without rewriting all tests etc pp. Using grunt and karma-chrome-launcher sounds good, but I havent any experience with it, do you? How much rework would that be?

**Comment by niksy** [comment]
@rejas I would definitely go the route of setting up tests with minimal architecture changes, so only browser tests in Karma for now and then we will see for anything else later.

I have experience with Karma (on daily basis I would say :D) and I can set up test environment with Puppeteer. I suppose it would be too much work to do that since Mocha is browser friendly, but it remains to be seen :)

I don’t know when this will be ready since I’m going on vacation for 2 weeks, but in the meantime if anything changes with current implementation, great! If not, I will try to create minimal PoC with Karma.

**Comment by rejas** [comment]
Have a nice vacation, those test issue probably wont go anywhere (it works somehow :-)

**Comment by rejas** [comment]
Will read a little about testing in the meantime, will need it for other projects (and work) too. 

Maybe I pick up work on the update_chai branch again where I only try to replace the old expect.js lib with chai.js. Still two strange errors in it though, not sure where they come from (maybe mocha?) So, yeah that might be a starting point for you too?

Will need a node and brwoser based environment as far as I understand the grunt / code correctly.

And what do you think about jest by the way?

**Comment by niksy** [comment]
So… I actually had some time and I whipped up initial conversion to Karma runner :D

You can see it here: https://github.com/niksy/Modernizr/commit/b58fec2d0eaa7a33d91f7cce80399ddcd0111f57

Every test passes in Chrome, so testing in other browsers is a process of setting up additional launchers and setting up Sauce Labs support.

This still needs coverage reporting and integration tests but we’re getting somewhere :D

As for Jest… I don’t have any experience with Node testing frameworks other that Mocha, and I think Jest still doesn’t have proper *real* browser support, and I think Modernizr needs that.

So, what do you think? And @patrickkettner, what do you think about switch to Karma runner for tests?

**Comment by patrickkettner** [comment]
Open to it, historically we didn't because it crashed in older browsers which is one of the more popular situations to use modernizr

**Comment by niksy** [comment]
Ah, I see. So maybe we should still keep original implementation, but fix hanging tests?

**Comment by escodel** [comment]
Not sure if this is related, but I saw this issue was still open and I'm now getting the following in my mocha:test task when running `grunt browserResults` 

``` 
(node:10272) [DEP0013] DeprecationWarning: Calling an asynchronous function without callback is deprecated.
Testing: http://localhost:9999/test/index.html


  ......
Warning: PhantomJS timed out, possibly due to a missing Mocha run() call. Use --force to continue.
```

Tried updating grunt-mocha but got way more errors/warnings so reverted back to what's currently in use. Also tried installing grunt-mocha-phantomjs to no avail.

Node 8.12.0
npm 6.4.1

Any known way to resolve this? Thanks in advance!

**Comment by rejas** [comment]
@escodel I have the same versions and can run `grunt browserResults` as well as `npm test` without hang-ups. The DepreactionWarnings shouldnt be a problem (yet) but no time to check those. 
I am running on Win10, you too?

**Comment by escodel** [comment]
@rejas Yep, Win10 for me too. I’m thinking another package might’ve updated the version/path for PhantomJS. Will check when I get time later today.

**Comment by rejas** [comment]
Weird, I have this issue only on my mac book....

**Comment by rejas** [comment]
(and yeah, upgrading grunt/mocha leads to another path full of errors, that bug is not for the faint of heart))

**Comment by escodel** [comment]
Alright so I figured out the issue... kind of. `navigator.connection` was returning undefined causing PhantomJS to get hung up. The other feature-detect for connection has a polyfill but I think this is an issue with the version of phantom not supporting navigator [bug](https://github.com/ariya/phantomjs/issues/10647). Will try and update phantom again.

**Comment by rejas** [comment]
Any news on updating phantom @escodel ?

**Comment by escodel** [comment]
@rejas I think it was an installation error on my part. Conflicting versions. It worked once I un/re-installed.

**Comment by rejas** [comment]
@escodel darn, I was hoping to find someone who knows a fix to the tests dont run on node10 and some other setups.... :-(

**Comment by rejas** [comment]
Fixed via #2429
